### PR TITLE
SP2-148 refractor result holder

### PR DESCRIPTION
**Pull request by frske**
This has been a ride, very emotional too.

The solution we came up with after discussing it also in the team was simply creating a new Model only for the formatted results. To format the results there is a nested foreach, I am very sorry about that but should not be bad overall as we expect the nested lists to not be big (under 10 elements). 

I still look very much forward to your feedback and review.

### CODE REVIEW

File: Heatington/ResultDataManager/ResultDataManager.cs [hunk 1]
```diff
@@ -1,21 +1,56 @@
+using System.Buffers;
 using Heatington;
 using Heatington.Controllers;
 using Heatington.Models;
 using Heatington.Optimizer;
 
 namespace Heatington.ResultDataManager;
 
-public class ResultDataManager(CsvController _csvController)
+public class ResultDataManager
```
Comment (magwrap) [inline Heatington/ResultDataManager/ResultDataManager.cs:9]: so no dependency injection?

File: Heatington/Controllers/CsvController.cs
Comment (magwrap) [inline Heatington/Controllers/CsvController.cs:1]: this is reverted change right? Like this code existed before?

File: .gitignore [hunk 2]
```diff
@@ -949,3 +949,4 @@ $RECYCLE.BIN/
 # End of https://www.toptal.com/developers/gitignore/api/csharp,dotnetcore,aspnetcore,visualstudio,visualstudiocode,rider,emacs,vim,macos,windows,linux
 Heatington.sln.DotSettings
 # Heatington/Program.cs
+Heatington.Console/Program.cs
```
Comment (magwrap) [inline .gitignore:952]: Finally

**Review by magwrap — COMMENTED** [review]
My only question is about the RDM - do we inject the csv controller or is it defined inside of the class?

File: Heatington/ResultDataManager/ResultDataManager.cs (hunk 1 above)
Comment (pkyeager) [reply Heatington/ResultDataManager/ResultDataManager.cs:9]: When I am still updated we want to go singleton design pattern and full microservice with loose coupling. 

Therefore I assume that Francesco's solution with having only the constructor with 1 param and without contructor injection which is in this case filePath and the instantiation of the CsvController is preferable.

File: Heatington/Models/FormatedResultHolder.cs [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    ProductionUnit boiler,
+    double netProductionCost)
+{
+    public DateTime StartTime { get; } = startTime;
+    public DateTime EndTime { get; } = endTime;
+    public double HeatDemand { get; } = heatDemand;
+    public double ElectricityPrice { get; } = electricityPrice;
+    public ProductionUnit Boilers { get; set; } = boiler;
+    public double NetProductionCost { get; set; } = netProductionCost;
+}
```
Comment (pkyeager) [inline Heatington/Models/FormatedResultHolder.cs:20]: I am not sure, but the FormattedResultHolder props will get set for only in the population phase?

If so consider using init?
https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/init
```suggestion

public class FormatedResultHolder(
    DateTime startTime,
    DateTime endTime,
    double heatDemand,
    double electricityPrice,
    ProductionUnit boiler,
    double netProductionCost)
{
    public DateTime StartTime { get; } = startTime;
    public DateTime EndTime { get; } = endTime;
    public double HeatDemand { get; } = heatDemand;
    public double ElectricityPrice { get; } = electricityPrice;
    public ProductionUnit Boilers { get; init; } = boiler;
    public double NetProductionCost { get; init; } = netProductionCost;
}
```

File: Heatington/Controllers/CsvController.cs [hunk 4]
```diff
@@ -59,5 +59,22 @@ public void SaveData(List<ResultHolder> data)
...
+                _fileController.WriteData<string>(rawCsvData);
+            }
+            catch (Exception e)
+            {
+                Utilities.DisplayException(e.Message);
+                throw;
+            }
+        }
     }
 }
```
Comment (pkyeager) [inline Heatington/Controllers/CsvController.cs:80]: I am a bad coder and I also have no clue of computer science or programming in general, but before merging this, shouldn't we maybe try this out? It was already in the todo and this should be a quick change?
```suggestion
public void SaveData<T>(List<T> data)
{
    try
    {
        CsvData csv = CsvData.Create(data);
        string rawCsvData = CsvSerializer.Serialize(csv);
        _fileController.WriteData<string>(rawCsvData);
    }
    catch (Exception e)
    {
        Utilities.DisplayException(e.Message);
        throw;
    }
}
```

File: Heatington.Console/ConsoleUI.cs [hunk 5]
```diff
@@ -93,7 +93,7 @@ private void LoadData()
         _dataPoints = sdm.TimeSeriesData!;
 
         opt.LoadData();
-        opt.OptimizeScenario1();
+        opt.Optimize();
```
Comment (pkyeager) [inline Heatington.Console/ConsoleUI.cs:96]: I don't see any optimize method from the optimizer. I guess, that OPT is planned to have only 1 main optimize method then?

File: .gitignore (hunk 2 above)
Comment (pkyeager) [reply .gitignore:952]: !!

File: Heatington/Controllers/CsvController.cs (hunk 4 above)
Comment (pkyeager) [reply Heatington/Controllers/CsvController.cs:80]: PS: didn't test this.

**Comment by frske** [comment]
I do not know why both methods are still in there, I removed the optimizeScenario1 altogether

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by frske** [comment]
I didn't want to do that because never played with generics, other parts of the code rely on it and it's not code I really know right now.

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by frske** [comment]
Won't be using init, as it might get run more than once when the class is already instantiated and or optimization results change

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Review by KonyakB — APPROVED** [review]
lgtm

**Review by KonyakB — APPROVED** [review]
now that the tests passed i can rebase

**Pull request merged**